docs/internal/abi.md

```diff
@@ -0,0 +1,444 @@
+Some targets have stronger alignment requirements, and unaligned reads can incur in performance penalties.
+Having an 8-bit type between 64-bit types can misalign the larger objects.
+By sorting the types by size, we guarantee that all 64-bit values are 64-bit aligned, all 32-bit values are 32-bit aligned and so on.
+
+A more optimal layout for the same class above, on either 32 or 64 bit targets, would be:
+```ts
+ { { ptr, ptr }, f64,   ptr,  i32 }
+      header,    F64, Other*, U32
+```
+
+Because `f64` aligns stronger than `i32` and `ptr` can be either.
+On CHERI, the 128-bit `ptr` would be the first field after the header.
+
+### Dispatch Tables & Selector Colouring
+
+Object headers have a pointer to their type's dispatch table.
+A dispatch table contains pointers to the functions that the type provides, at specific offsets.
+Those pointers are called when it's not possible to determine the actual function being called at compile time.
+There is only one dispatch table per type, not per object.
+
+The offset of each method is calculated at compile time.
+The address at each offset has a pointer to the actual functions.
+Dynamic dispatch is done by taking the pointer at the offset from a table pointer and calling that.
+If an interface provides a method `foo`, all classes that implement that interface will have a method `foo` and their dispatch tables will have an entry for it.
+
+To avoid each type to have a different calculation for each method's offset, the compiler will do **selector colouring**.
+This process finds all common methods across all types and ensure the same methods all have the same offsets on all tables.
+This applies to concrete classes that implement specific interfaces, but it's not limited by it.
+Any two classes that have the same methods (signature) will end up with the same offset.
+
+With multiple interfaces and classes implementing similar methods, this can lead to a number of gaps between offsets.
+The tables will be compacted (either during colouring or afterwards) to minimise that.
+
+## Union Types
+
+Union types need as many representations as there are types in the union.
+Objects of union types need to be able to represent all types as well as identify which one for each instance.
+
+Example:
+```ts
+// Must handle both U8 and F32
+// Must also know which one at run time
+function(arg : U8 | F32) { ... }
+...
+var a : U8 & imm = 10;
+function(a); // Passes a 1-byte integer
+var b : F32 & imm = 3.14;
+function(b); // Passes 32-bit floating point
+```
+
+All objects are referred to by pointers.
+Machine-word types are refereed to by value as an optimisation by the compiler.
+
+For that reason, they can only _contain_ the following representations:
+* Machine-word values.
+* Pointers to objects.
+
+To identify which type the run-time object has, we need a discriminator flag.
+Because the values take up all their storage, we need a wider storage to bundle more than one type in the same representation.
```

Comment:
```suggestion
Because values generally take up all their storage, we usually need storage wider than the largest union member to bundle more than one type in the same representation.
```